python/cudf/cudf/core/spillable_buffer.py

```diff
@@ -0,0 +1,445 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import collections.abc
+import pickle
+import time
+import weakref
+from threading import RLock
+from typing import TYPE_CHECKING, Any, Dict, List, Tuple
+
+import numpy
+
+import rmm
+
+from cudf.core.buffer import Buffer, Frame, get_ptr_and_size
+from cudf.utils.string import format_bytes
+
+if TYPE_CHECKING:
+    from cudf.core.spill_manager import SpillManager
+
+
+class SpillLock:
+    pass
+
+
+class DelayedPointerTuple(collections.abc.Sequence):
+    """
+    A delayed version of the "data" field in __cuda_array_interface__.
+
+    The idea is to delay the access to `Buffer.ptr` until the user
+    actually accesses the data pointer.
+
+    For instance, in many cases __cuda_array_interface__ is accessed
+    only to determine whether an object is a CUDA object or not.
+
+    TODO: this doesn't support libraries such as PyTorch that declare
+    the tuple of __cuda_array_interface__["data"] in Cython. In such
+    cases, Cython will raise an error because DelayedPointerTuple
+    isn't a "real" tuple.
+    """
+
+    def __init__(self, buffer) -> None:
+        self._buf = buffer
+
+    def __len__(self):
+        return 2
+
+    def __getitem__(self, i):
+        if i == 0:
+            return self._buf.ptr
+        elif i == 1:
+            return False
+        raise IndexError("tuple index out of range")
+
+
+class SpillableBuffer(Buffer):
+    """A spillable buffer that implements DeviceBufferLike.
+
+    This buffer supports spilling the represented data to host memory.
+    Spilling can be done manually by calling `.__spill__(target="cpu")`
+    but usually the associated spilling manager triggers spilling based on
+    current device memory usage see `cudf.core.spill_manager.SpillManager`.
+    Unspill is triggered automatically when accessing the data of the buffer.
+
+    The buffer might not be spillable, which is based on the "expose" status
+    of the buffer. We say that the buffer has been exposed if the device
+    pointer (integer or void*) has been accessed outside of SpillableBuffer.
+    In this case, we cannot invalidate the device pointer by moving the data
+    to host.
+
+    A buffer can be exposed permanently at creation or by accessing the `.ptr`
+    property. To avoid this, one can use `.get_ptr()` instead, which support
+    exposing the buffer temporarily.
+
+    Parameters
+    ----------
+    data : buffer-like
+        An buffer-like object representing device or host memory.
+    exposed : bool, optional
+        Whether or not a raw pointer (integer or C pointer) has
+        been exposed to the outside world. If this is the case,
+        the buffer cannot be spilled.
+    manager : SpillManager
+        The manager overseeing this buffer.
+    """
+
+    _ptr_desc: Dict[str, Any]
+    _spill_locks: weakref.WeakSet
+
+    def __init__(
```

Comment:
This constructor diverges from the parent constructor. Is there no situation where we would need to construct a `SpillableBuffer` directly from a pointer/size combination?

Reply:
Fixed in 660316588ddc1d99cca0b74391fcd8fc0f84e14d